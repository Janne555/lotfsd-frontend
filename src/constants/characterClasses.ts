const characterClasses = {
  fighter: {
    savingThrows: {
      breathWeapon: 10,
      magic: 11,
      magicalDevice: 12,
      paralyze: 13,
      poison: 14
    },
    levels: [
      0, 2000, 4000, 8000, 16000, 32000, 64000
    ]
  }
}

export { characterClasses }