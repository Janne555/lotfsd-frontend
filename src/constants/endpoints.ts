const APIROOT = process.env.NODE_ENV === 'production' ? "https://lotfsdapi.azurewebsites.net/" : "http://172.17.0.2:5000/"

const ENDPOINTS = {
  LOGIN: { url: 'user/login', type: "" },
  REGISTER: { url: 'user/register', type: "" }
}

export {
  APIROOT,
  ENDPOINTS
}