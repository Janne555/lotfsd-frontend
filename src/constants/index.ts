const ATTRIBUTE_DETAILS = {
  charisma: "Retainer Recruitment, Loyalty",
  constitution: "Hit Points, Daily Travel Distance",
  dexterity: "AC, Ranged AB, Initiative",
  intelligence: "Saves vs Magic Effects",
  strength: "Mêlée AB, Open Doors",
  wisdom: "Saves vs Non-Magic Effects"
}

const SAVING_THROW_DETAILS = {
  paralyze: "Mobility Hazards (Petrification, Hold, Etc.)",
  poison: "Instant Death/KO Situations",
  breathWeapon: "Area Effects",
  magicalDevice: "Spell-Like Effects from Items",
  magic: "Spells or Innate Abilities"
}

const ATTRIBUTE_TITLES = {
  charisma: "Charisma",
  constitution: "Constitution",
  dexterity: "Dexterity",
  intelligence: "Intelligence",
  strength: "Strength",
  wisdom: "Wisdom"
}

const SAVING_THROW_TITLES = {
  paralyze: "Paralyze",
  poison: "Poison",
  breathWeapon: "Breath\xa0Weapon",
  magicalDevice: "Magical\xa0Device",
  magic: "Magic"
}

const COMMON_ACTIVITY_TITLES = {
  architecture: "Architecture",
  bushcraft: "Bushcraft",
  climbing: "Climbing",
  languages: "Languages",
  openDoors: "Open Doors",
  search: "Search",
  sleightOfHand: "Sleight of Hand",
  sneakAttack: "Sneak Attack",
  stealth: "Stealth",
  tinkering: "Tinkering"
}

const TITLES = {
  BASE_AB: "Base AB",
  MELEE_AB: "Mêlée AB",
  RANGED_AB: "Ranged AB"
}

const BASE_ARMOR_CLASS = 12

const MONEY = 'Money'

const RETAINER_KEYS: (keyof Omit<Retainer, 'uuid'>)[] = [
  "name",
  "position",
  "class",
  "level",
  "hitpoints",
  "wage",
  "share"
]

export {
  ATTRIBUTE_DETAILS,
  SAVING_THROW_DETAILS,
  ATTRIBUTE_TITLES,
  SAVING_THROW_TITLES,
  TITLES,
  BASE_ARMOR_CLASS,
  COMMON_ACTIVITY_TITLES,
  MONEY,
  RETAINER_KEYS
}