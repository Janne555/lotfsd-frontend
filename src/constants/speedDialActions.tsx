import React, { ReactNode } from 'react'
import { Icon } from '@material-ui/core'
import { match as Match } from 'react-router-dom'

type SpeedDialAction = {
  action: string | ((root: string) => string)
  icon: ReactNode | undefined
  tooltip: string
}

const speedDialActions: Record<string, SpeedDialAction[]> = {
  characters: [
    { action: root => `/characters/${root}/addItem`, icon: <Icon />, tooltip: 'Add\xa0Item' },
    { action: root => `/characters/${root}/addRetainer`, icon: <Icon />, tooltip: 'Add\xa0Retainer' },
    { action: root => `/characters/${root}/addLanguage`, icon: <Icon />, tooltip: 'Add\xa0Language' },
  ],
  itemindex: [
    { action: '/newitem', icon: <Icon />, tooltip: 'New\xa0Item' },
  ]
}

function getSpeedDialActions(match: Match<{ page: string }> | null): SpeedDialAction[] {
  const page = match?.params.page
  if (page && speedDialActions[page]) {
    return speedDialActions[page]
  } else {
    return []
  }
}


export { getSpeedDialActions }