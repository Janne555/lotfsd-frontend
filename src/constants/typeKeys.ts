const COMMON_ACTIVITY_KEYS = [
  "architecture",
  "bushcraft",
  "climbing",
  "languages",
  "openDoors",
  "search",
  "sleightOfHand",
  "sneakAttack",
  "stealth",
  "tinkering"
] as const

const ATTRIBUTE_KEYS = [
  'charisma',
  'constitution',
  'dexterity',
  'intelligence',
  'strength',
  'wisdom'
] as const


const RETAINER_KEYS = [
  "name",
  "position",
  "class",
  "level",
  "hitpoints",
  "wage",
  "share"
] as const


export {
  COMMON_ACTIVITY_KEYS,
  ATTRIBUTE_KEYS,
  RETAINER_KEYS
}

