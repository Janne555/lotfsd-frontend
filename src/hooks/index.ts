import { useDispatch, useSelector } from './redux'
import { useState, useLayoutEffect } from 'react'

function useScreenResizeEvent(condition: (screenWidth: number) => boolean) {
  const [result, setResult] = useState(condition(window.screen.width))
  useLayoutEffect(() => {
    function listener() {
      const now = condition(window.screen.width)
      if (now !== result) {
        setResult(now)
      }
    }
    window.addEventListener('resize', listener)
    return () => window.removeEventListener('resize', listener)
  })
  return result
}

export {
  useDispatch,
  useSelector,
  useScreenResizeEvent
}