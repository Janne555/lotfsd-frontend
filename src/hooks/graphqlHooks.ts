import { useMutation } from '@apollo/react-hooks'
import { CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables } from '../../__generated__/apolloTypes/CharacterSheetUpdateMutation'
import { CHARACTER_SHEET_UPDATE_MUTATION } from '../constants'
import { updateCharacterSheet } from '../services'

function useCharacterSheetUpdate(characterId?: string) {
  return useMutation<CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables>(CHARACTER_SHEET_UPDATE_MUTATION, {
    update: updateCharacterSheet(characterId),
  })
}

export {
  useCharacterSheetUpdate
}