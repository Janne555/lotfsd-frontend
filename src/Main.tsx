import React from 'react'
import CharacterSheet from './Components/CharacterSheet/CharacterSheet'
import { createUseStyles } from 'react-jss'
import NavBar from './Components/Interface/NavMenu/NavBar'
import NavItem from './Components/Interface/NavMenu/NavItem'
import { Switch, Route, useRouteMatch, Redirect } from 'react-router-dom'
import CharacterList from './Components/Interface/CharacterList/CharacterList'
import CharacterCreator from './Components/CharacterCreator/CharacterCreator'
import AppModal from './Components/Interface/Modal/AppModal'
import ItemIndex from './Components/ItemIndex/ItemIndex'
import ItemCreator from './Components/ItemIndex/ItemCreator'
import { useLogin, useScreenResizeEvent } from './hooks'
import { useQuery } from '@apollo/react-hooks'
import { CHARACTER_LIST_QUERY, ITEMS_QUERY } from './constants'
import { CharacterListQuery } from '../__generated__/apolloTypes/CharacterListQuery'
import SpeedDial from './Components/Interface/SpeedDial'
import CircularProgress from '@material-ui/core/CircularProgress'
import { ItemsQuery } from '../__generated__/apolloTypes/ItemsQuery'

const useStyles = createUseStyles((theme: Theme) => ({
  app: {
    display: 'grid',
    gridTemplateColumns: 'auto 1100px auto',
    gridTemplateRows: '3rem auto 5rem',
    justifyContent: 'center',
    '& > *': {
      gridColumnStart: 2
    }
  },
  footer: {
    backgroundColor: 'pink'
  },
  body: {
    minHeight: 'calc(100vh - 8rem)',
    backgroundColor: theme.colorGreyLightest,
  },
  error: {
    position: 'absolute',
    zIndex: 1000,
    height: '100vh',
    width: '100vw',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center'
  },
  '@media (max-width: 1100px)': {
    app: {
      display: 'flex',
      flexDirection: 'column'
    }
  }
}))

function Main() {
  const classes = useStyles()
  const characterNameMatch = useRouteMatch<{ character: string }>('/characters/:character')
  const actionMatch = useRouteMatch<{ action: string }>('/characters/:character/:action')
  const { data, loading, error } = useQuery<CharacterListQuery>(CHARACTER_LIST_QUERY)
  const { data: data2, loading: loading2, error: error2 } = useQuery<ItemsQuery>(ITEMS_QUERY)
  const { logout } = useLogin()
  const isMobile = useScreenResizeEvent(width => width < 1100)

  if (loading || loading2) {
    return <CircularProgress />
  }

  if (error || error2) {
    return (
      <div className={classes.error}>
        <h3>Failed to connect to server</h3>
      </div>
    )
  }

  if (!data || !data.characterSheets || !data2 || !data2.items) {
    return null
  }

  const characterId = data.characterSheets.find(ch => ch.name === characterNameMatch?.params.character)?.id


  return (
    <div className={classes.app}>
      <NavBar>
        <NavItem name="Characters" to="/characters">
          <CharacterList characters={data.characterSheets} />
        </NavItem>
        <NavItem name="Campaigns" to="/campaigns">
          moi
          </NavItem>
        <NavItem name="Item Index" to="/itemindex" />
        <NavItem name="Logout" end onClick={logout} />
      </NavBar>
      <div className={classes.body}>
        <Switch>
          <Route path="/characters/:character">
            {characterId && characterNameMatch?.params.character &&
              <CharacterSheet characterId={characterId} characterName={characterNameMatch?.params.character} itemIndex={data2.items} />
            }
          </Route>
          <Route exact path="/characters">
            <CharacterList characters={data.characterSheets} />
          </Route>
          <Route path="/newcharacter">
            <CharacterCreator />
          </Route>
        </Switch>
        <Route path="/itemindex">
          <ItemIndex />
        </Route>
        <Route path="/newitem">
          <ItemCreator />
        </Route>
        <Route exact path="/">
          <Redirect to="/characters" />
        </Route>
        <Route exact path="/login">
          <Redirect to="/characters" />
        </Route>
      </div>

      <AppModal characterId={characterId} content={actionMatch?.params.action} characterName={characterNameMatch?.params.character} />

      <footer className={classes.footer}></footer>

      {isMobile &&
        <SpeedDial />
      }
    </div>
  );
}

export default Main