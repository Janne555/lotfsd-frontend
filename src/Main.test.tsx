import React, { ReactNode } from 'react'
import { JssProvider } from 'react-jss'
import { render, waitForDomChange } from '@testing-library/react'
import { MockedProvider, MockedResponse } from '@apollo/react-testing'
import { MemoryRouter } from 'react-router-dom'
import { CHARACTER_LIST_QUERY, ITEMS_QUERY } from './constants'
import Main from './Main'
import { ItemsQuery } from '../__generated__/apolloTypes/ItemsQuery'
import { CharacterListQuery } from '../__generated__/apolloTypes/CharacterListQuery'
import { LoginProvider } from './hooks'

type Result<T> = {
  data: T
}

const CHARACTER_LIST_MOCK: Result<CharacterListQuery> = {
  data: {
    characterSheets: [
      {
        __typename: "CharacterSheet",
        class: "fighter",
        id: "213123",
        name: "zoid"
      }
    ]
  }
}

const ITEM_INDEX_MOCK: Result<ItemsQuery> = {
  data: {
    items: [
      {
        __typename: "Item",
        attackBonus: null,
        baseArmorClass: null,
        damage: null,
        description: "a description",
        effects: null,
        encumbrance: null,
        encumbrancePoints: 1,
        id: "asd",
        name: "thingy",
        rangeLong: null,
        rangeMedium: null,
        rangeShort: null,
        stackSize: 1,
        type: "item"
      }
    ]
  }
}

const mocks: MockedResponse[] = [
  {
    request: {
      query: CHARACTER_LIST_QUERY
    },
    result: CHARACTER_LIST_MOCK
  },
  {
    request: {
      query: ITEMS_QUERY
    },
    result: ITEM_INDEX_MOCK
  }
]

describe('<Main />', () => {
  it('should match snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginProvider>
          <MockedProvider mocks={mocks}>
            <JssProvider disableStylesGeneration>
              <Main />
            </JssProvider>
          </MockedProvider>
        </LoginProvider>
      </MemoryRouter>
    )

    await waitForDomChange()

    expect(asFragment()).toMatchSnapshot()
  });
});