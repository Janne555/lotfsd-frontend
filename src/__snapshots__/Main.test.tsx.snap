// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div>
        <h2>
          Characters
        </h2>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              color="white"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <div>
      <div>
        <div>
          <button>
            <img
              alt="class"
              src="/images/fighter.svg"
            />
            <h2>
              zoid
            </h2>
          </button>
          <button>
            <h2>
              Create new
            </h2>
          </button>
        </div>
      </div>
    </div>
    <footer />
    <div
      class="MuiSpeedDial-root MuiSpeedDial-directionUp"
      role="presentation"
    >
      <button
        aria-controls="-actions"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label=""
        class="MuiButtonBase-root MuiFab-root MuiSpeedDial-fab MuiFab-primary"
        style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        aria-orientation="vertical"
        class="MuiSpeedDial-actions MuiSpeedDial-actionsClosed"
        id="-actions"
        role="menu"
      >
        <span
          class="MuiSpeedDialAction-staticTooltip MuiSpeedDialAction-tooltipPlacementLeft MuiSpeedDialAction-staticTooltipClosed"
          id="-action-0"
        >
          <span
            class="MuiSpeedDialAction-staticTooltipLabel"
            id="-action-0-label"
            style="transition-delay: 90ms;"
          >
            Add Item
          </span>
          <button
            aria-describedby="-action-0-label"
            class="MuiButtonBase-root MuiFab-root MuiSpeedDialAction-fab MuiSpeedDialAction-fabClosed MuiFab-sizeSmall"
            role="menuitem"
            style="transition-delay: 90ms;"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <span
                aria-hidden="true"
                class="material-icons MuiIcon-root"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
        <span
          class="MuiSpeedDialAction-staticTooltip MuiSpeedDialAction-tooltipPlacementLeft MuiSpeedDialAction-staticTooltipClosed"
          id="-action-1"
        >
          <span
            class="MuiSpeedDialAction-staticTooltipLabel"
            id="-action-1-label"
            style="transition-delay: 60ms;"
          >
            Add Retainer
          </span>
          <button
            aria-describedby="-action-1-label"
            class="MuiButtonBase-root MuiFab-root MuiSpeedDialAction-fab MuiSpeedDialAction-fabClosed MuiFab-sizeSmall"
            role="menuitem"
            style="transition-delay: 60ms;"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <span
                aria-hidden="true"
                class="material-icons MuiIcon-root"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
        <span
          class="MuiSpeedDialAction-staticTooltip MuiSpeedDialAction-tooltipPlacementLeft MuiSpeedDialAction-staticTooltipClosed"
          id="-action-2"
        >
          <span
            class="MuiSpeedDialAction-staticTooltipLabel"
            id="-action-2-label"
            style="transition-delay: 30ms;"
          >
            Add Language
          </span>
          <button
            aria-describedby="-action-2-label"
            class="MuiButtonBase-root MuiFab-root MuiSpeedDialAction-fab MuiSpeedDialAction-fabClosed MuiFab-sizeSmall"
            role="menuitem"
            style="transition-delay: 30ms;"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <span
                aria-hidden="true"
                class="material-icons MuiIcon-root"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
